webapp/src/ts/services/create-user-for-contacts.service.ts

```diff
@@ -0,0 +1,112 @@
+import { Injectable } from '@angular/core';
+import { DbService } from '@mm-services/db.service';
+import { DBSyncService, SyncStatus } from '@mm-services/db-sync.service';
+import { SessionService } from '@mm-services/session.service';
+import { SettingsService } from '@mm-services/settings.service';
+import { UserSettingsService } from '@mm-services/user-settings.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CreateUserForContactsService {
+  constructor(
+    private settingsService: SettingsService,
+    private userSettingsService: UserSettingsService,
+    private dbService: DbService,
+    private dbSyncService: DBSyncService,
+    private sessionService: SessionService,
+  ) {
+    const that = this;
+    this.settingsService
+      .get()
+      .then((settings) => {
+        if (settings?.transitions?.create_user_for_contacts) {
+          that.dbSyncService.subscribe(async(status) => {
+            return that.syncStatusChanged(status);
+          });
+        }
+      });
+  }
+
+  async getUserContact() {
+    const { contact_id }: any = await this.userSettingsService.get();
+    if (!contact_id) {
+      return;
+    }
+    return this.dbService
+      .get()
+      .get(contact_id)
+      .catch(err => {
+        if (err.status === 404) {
+          return;
+        }
+        throw err;
+      });
+  }
+
+  setReplaced(originalContact, newContact) {
+    if (!originalContact) {
+      throw new Error('The original contact could not be found when replacing the user.');
+    }
+    if (!newContact) {
+      throw new Error('The new contact could not be found when replacing the user.');
+    }
+    if (!originalContact.parent?._id || originalContact.parent._id !== newContact.parent._id) {
+      throw new Error('The new contact must have the same parent as the original contact when replacing a user.');
+    }
+    const status = this.sessionService.isOnlineOnly() ? UserCreationStatus.READY : UserCreationStatus.PENDING;
+    if(!originalContact.user_for_contact) {
+      originalContact.user_for_contact = {};
+    }
+    originalContact.user_for_contact.replaced = { status, by: newContact._id };
+  }
+
+  isReplaced(contact) {
+    return !!contact.user_for_contact?.replaced;
+  }
+
+  getReplacedBy(contact): string {
+    return contact.user_for_contact?.replaced?.by;
+  }
+
+  private getReplacedStatus(contact): UserCreationStatus {
+    return UserCreationStatus[contact.user_for_contact.replaced.status];
+  }
+
+  private setReplacedStatus(contact, status: UserCreationStatus) {
+    contact.user_for_contact.replaced.status = status;
+  }
+
+  private async syncStatusChanged({ to, from }: any) {
+    if (to !== SyncStatus.Success || from !== SyncStatus.Success) {
```

Comment:
I have been able to confirm that a success status is pushed when a `denied` event is emitted on replication (updated the unit test to be explicitly clear about this). I checked the [Pouch Docs](https://pouchdb.com/api.html#replication) for what triggers the `denied` event: 

> This event fires if a document failed to replicate due to validation or authorization errors.

Contextually, it seems like these events occur when trying to replicate a doc that violates our business rules in API (and not when just any random `4xx` status comes back from the server).  So, this would happen if a user tries to create a document associated with a person that he actually does not have access to and stuff like that.  But, this is not the event that gets triggered if the user is halfway through a sync and looses internet connectivity....  

All that is to say, I wonder if perhaps we are okay with how things work now. In theory, if replication is denied for a particular doc (and all else remains the same) that doc will _never_ be allowed to replicate right?  I am thinking maybe we do not care to "lose" docs that already would never be replicated to the server....  That being said, I realize data loss is _extremely important_ to avoid, so let me know if you think we need a safer method of ensuring the data all is synced!  (I can look into your checkpointer method except you seemed to imply that is not perfect either...?)

Reply:
The denied document can become valid for replication, which is why we have a retry mechanism. 
Some background around that: https://github.com/medic/cht-core/issues/6103

Reply:
For the record, I discussed this more with @dianabarsan.  The conclusion was that the issue of the sync status being `success` even when docs are `denied` is a larger issue than this project and should be addressed on its own in.  https://github.com/medic/cht-core/issues/7843

We believe that the current implementation we have for this project is sufficient for now.  (And once #7842 is addressed, the current code should be able to take advantage of the enhancement without needing additional changes!)